File: src/main/java/editor/filter/leaf/ManaCostFilter.java
package editor.filter.leaf;

import java.util.Objects;

import com.google.gson.JsonObject;

import editor.database.attributes.CardAttribute;
import editor.database.attributes.ManaCost;
import editor.database.card.Card;
import editor.util.Containment;

/**
 * This class represents a filter to group cards by mana costs.
 *
 * @author Alec Roelke
 */
public class ManaCostFilter extends FilterLeaf<ManaCost>
{
    /**
     * Containment for this ManaCostFilter.
     */
    public Containment contain;
    /**
     * Mana cost to filter by.
     */
    public ManaCost cost;

    /**
     * Create a new ManaCostFilter.
     */
    public ManaCostFilter()
    {
        super(CardAttribute.MANA_COST, null);
        contain = Containment.CONTAINS_ANY_OF;
        cost = new ManaCost();
    }

    @Override
    public FilterLeaf<ManaCost> copyLeaf()
    {
        ManaCostFilter filter = (ManaCostFilter)CardAttribute.createFilter(CardAttribute.MANA_COST);
        filter.contain = contain;
        filter.cost = cost;
        return filter;
    }

    @Override
    public boolean leafEquals(Object other)
    {
        if (other.getClass() != getClass())
            return false;
        ManaCostFilter o = (ManaCostFilter)other;
        return o.contain == contain && o.cost.equals(cost);
    }

    @Override
    public int hashCode()
    {
        return Objects.hash(type(), contain, cost);
    }

    /**
     * {@inheritDoc}
     * Filter cards by their mana costs.
     */
    @Override
    public boolean testFace(Card c)
    {
        return switch (contain) {
            case CONTAINS_ANY_OF -> Containment.CONTAINS_ANY_OF.test(c.manaCost(), cost);
            case CONTAINS_NONE_OF -> Containment.CONTAINS_NONE_OF.test(c.manaCost(), cost);
            case CONTAINS_ALL_OF -> c.manaCost().isSuperset(cost);
            case CONTAINS_NOT_ALL_OF -> !c.manaCost().isSuperset(cost);
            case CONTAINS_EXACTLY -> c.manaCost().equals(cost);
            case CONTAINS_NOT_EXACTLY -> !c.manaCost().equals(cost);
        };
    }

    @Override
    protected void serializeLeaf(JsonObject fields)
    {
        fields.addProperty("contains", contain.toString());
        fields.addProperty("cost", cost.toString());
    }

    @Override
    protected void deserializeLeaf(JsonObject fields)
    {
        contain = Containment.parseContainment(fields.get("contains").getAsString());
        cost = ManaCost.parseManaCost(fields.get("cost").getAsString());
    }
}

File: src/main/scala/editor/filter/leaf/ManaCostFilter.scala
package editor.filter.leaf

import editor.database.attributes.CardAttribute
import editor.database.attributes.ManaCost
import editor.database.card.Card
import editor.filter.FaceSearchOptions
import editor.util.Containment
import editor.util.Containment._

/**
 * Filter that groups cards by mana cost.
 * 
 * @param contain function to use to compare with the card's mana cost
 * @param cost list of mana symbols to use for comparison
 * 
 * @author Alec Roelke
 */
final case class ManaCostFilter(faces: FaceSearchOptions = FaceSearchOptions.ANY, contain: Containment = AnyOf, cost: ManaCost = ManaCost()) extends FilterLeaf[ManaCostFilter] {
  override def attribute = CardAttribute.ManaCost
  override val unified = false
  override def testFace(c: Card) = contain match {
    case AnyOf      => AnyOf(c.manaCost, cost)
    case NoneOf     => NoneOf(c.manaCost, cost)
    case AllOf      => c.manaCost.isSuperset(cost)
    case SomeOf     => !c.manaCost.isSuperset(cost)
    case Exactly    => c.manaCost == cost
    case NotExactly => c.manaCost != cost
  }
  override def copyFaces(faces: FaceSearchOptions) = copy(faces = faces)
}